lib/srv/sessiontracker_test.go

```diff
@@ -27,8 +27,128 @@ import (
 
 	"github.com/gravitational/teleport/api/client/proto"
 	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
 )
 
+type failingTrackerService struct {
+	services.SessionTrackerService
+	updated     chan struct{}
+	updateError chan error
+}
+
+func (f *failingTrackerService) CreateSessionTracker(ctx context.Context, s types.SessionTracker) (types.SessionTracker, error) {
+	return s, nil
+}
+
+func (f *failingTrackerService) UpdateSessionTracker(ctx context.Context, req *proto.UpdateSessionTrackerRequest) error {
+	f.updated <- struct{}{}
+
+	return <-f.updateError
+}
+
+func TestSessionTracker_UpdateRetry(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	clock := clockwork.NewFakeClock()
+
+	svc := &failingTrackerService{
+		updated:     make(chan struct{}),
+		updateError: make(chan error),
+	}
+	spec := types.SessionTrackerSpecV1{
+		Created:   clock.Now(),
+		SessionID: "session",
+		Expires:   clock.Now().Add(10 * time.Hour),
+	}
+	tracker, err := NewSessionTracker(ctx, spec, svc)
+	require.NoError(t, err)
+	done := make(chan error)
+
+	updateError := trace.ConnectionProblem(context.DeadlineExceeded, "connection refused")
+
+	// start the update loop
+	go func() {
+		done <- tracker.updateExpirationLoop(ctx, clock)
+	}()
+
+	for i := 0; i < 10; i++ {
+		// wait for the ticker to be ready
+		if i%2 != 0 {
```

Comment:
this block-until logic is a bit opaque. i'm still not sure why on odd loops it waits for 2 sleepers and 1 on even loops. Something to do with the retry timer having been started I guess, but I dont see why. It seems like the code under test is either sleeping on the ticker channel or the retry channel but not both at once.

I am also wondering what the significance of doing this loop 10 times is. Would it make any difference if it were done 2 times? Does it test more code paths doing it more times? If not, can the whole thing be simplified to be more linear?

I am thinking of a helper function:
```
func waitForUpdate(t *testing.T, svc *failingTrackerService, done chan error) {
    t.Helper()
    select {
        case <-svc.updated:
        case err := <-done:
            t.Fatal("Update loop terminated early", err.Error())
    }
}
```
Then writing the test as a sequence of BlockUntil, Advance, waitForUpdate, send update error like:
```
clock.BlockUntil(1)
clock.Advance(sessionTrackerExpirationUpdateInterval)
waitForUpdate(t, svc, done)
svc.updateError <- updateError

clock.BlockUntil(1)
clock.Advance(45 * time.Second)
waitForUpdate(t, svc, done)
svc.updateError <- nil
...
```
Each section can be commented about what exactly it is driving in the code under test.

IMHO, tests need to be obviously right as much as they can be, and I am struggling with this one.